// react-templates/playground3/webpack.rt.config.js
/*eslint no-path-concat:0*/
'use strict';

var path = require('path');
var webpack = require('webpack');
var commonsPlugin = new webpack.optimize.CommonsChunkPlugin('commons.js');

module.exports = {
    //devtool: 'eval',
    devtool: 'source-map',
    context: __dirname,
    entry: {
        rt: ['./src/rt-main.js']
    },
    output: {
        filename: '[name].js',
        path: path.join(__dirname, 'dist'),
        publicPath: '/static/'
    },
    plugins: [
        new webpack.optimize.UglifyJsPlugin({
            compressor: {
                warnings: false
            }
        }),
        new webpack.optimize.DedupePlugin()
    ],
    externals: {
        // require("jquery") is external and available on the global var jQuery
        jquery: 'jQuery',
        react: 'React',
        'react/addons': 'React',
        lodash: '_'
    },
    resolve: {
        // you can now require('file') instead of require('file.coffee')
        extensions: ['', '.js', '.json'],
        alias: {},
    },
    module: {
        loaders: [
            {test: /\.js$/, exclude: /node_modules/, loaders: ['babel-loader']},
            {test: /\.json$/, loader: 'json-loader'},
            {test: /\.less$/, loader: 'style-loader!css-loader!less-loader'}, // use ! to chain loaders
            {test: /\.css$/, loader: 'style-loader!css-loader'},
            {test: /\.(png|jpg)$/, loader: 'url-loader?limit=8192'}, // inline base64 URLs for <=8k images, direct URLs for the rest
            {test: /\.rt/, exclude: /samples/, loaders: ['react-hot', 'react-templates-loader']},
            {test: /\.text/, loaders: ['raw-loader']}
            //"md|markdown": ["html-loader", "markdown-loader"]
        ]
    }
};
